pkg/policy/policy.go

```diff
@@ -0,0 +1,214 @@
+package policy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/open-policy-agent/opa/bundle"
+
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/aquasecurity/trivy/pkg/oci"
+
+	"golang.org/x/xerrors"
+	"k8s.io/utils/clock"
+)
+
+const (
+	bundleVersion    = 0 // Latest released MAJOR version for defsec
+	bundleRepository = "ghcr.io/aquasecurity/defsec"
+	policyMediaType  = "application/vnd.cncf.openpolicyagent.layer.v1.tar+gzip"
+	updateInterval   = 24 * time.Hour
+)
+
+type options struct {
+	artifact *oci.Artifact
+	clock    clock.Clock
+}
+
+// WithOCIArtifact takes an OCI artifact
+func WithOCIArtifact(art *oci.Artifact) Option {
+	return func(opts *options) {
+		opts.artifact = art
+	}
+}
+
+// WithClock takes a clock
+func WithClock(clock clock.Clock) Option {
+	return func(opts *options) {
+		opts.clock = clock
+	}
+}
+
+// Option is a functional option
+type Option func(*options)
+
+// Client implements policy operations
+type Client struct {
+	*options
+	policyDir string
+	quiet     bool
+	insecure  bool
+}
+
+// Metadata holds default policy metadata
+type Metadata struct {
+	Digest       string
+	DownloadedAt time.Time
+}
+
+// NewClient is the factory method for policy client
+func NewClient(cacheDir string, quiet bool, opts ...Option) (*Client, error) {
+	o := &options{
+		clock: clock.RealClock{},
+	}
+
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	return &Client{
+		options:   o,
+		policyDir: filepath.Join(cacheDir, "policy"),
+		quiet:     quiet,
+	}, nil
+}
+
+func (c *Client) populateOCIArtifact() error {
+	if c.artifact == nil {
+		repo := fmt.Sprintf("%s:%d", bundleRepository, bundleVersion)
+		art, err := oci.NewArtifact(repo, policyMediaType, c.quiet, c.insecure)
+		if err != nil {
+			return xerrors.Errorf("OCI artifact error: %w", err)
+		}
+		c.artifact = art
+	}
+	return nil
+}
+
+// DownloadBuiltinPolicies download default policies from GitHub Pages
+func (c *Client) DownloadBuiltinPolicies(ctx context.Context) error {
+	if err := c.populateOCIArtifact(); err != nil {
+		return xerrors.Errorf("OPA bundle error: %w", err)
+	}
+
+	dst := c.contentDir()
+	if err := c.artifact.Download(ctx, dst); err != nil {
+		return xerrors.Errorf("download error: %w", err)
+	}
+
+	digest, err := c.artifact.Digest()
+	if err != nil {
+		return xerrors.Errorf("digest error: %w", err)
+	}
+	log.Logger.Debugf("Digest of the built-in policies: %s", digest)
+
+	// Update metadata.json with the new digest and the current date
+	if err = c.updateMetadata(digest, c.clock.Now()); err != nil {
+		return xerrors.Errorf("unable to update the policy metadata: %w", err)
+	}
+
+	return nil
+}
+
+// LoadBuiltinPolicies loads default policies
+func (c *Client) LoadBuiltinPolicies() ([]string, error) {
+	f, err := os.Open(c.manifestPath())
```

Comment:
We need to close the file. Otherwise, it fails tests on Windows.

Reply:
ah... thanks for catching this!